### PR TITLE
[Android] Make sure referrer is received prior to service initialization

### PR DESCRIPTION
**Pull request by samartnik**
Resolves https://github.com/brave/brave-browser/issues/16031

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by AlexeyBarabash — APPROVED** [review]
++

**Comment by samartnik** [comment]
Audit failed due to reasons not related to this PR
```
Result: Audit failed due to vulnerabilities

[2021-05-27T17:18:21.972Z] {

[2021-05-27T17:18:21.972Z]   "1747": {

[2021-05-27T17:18:21.972Z]     "updated": "2021-05-24T19:59:05.419Z", 

[2021-05-27T17:18:21.972Z]     "severity": "moderate", 

[2021-05-27T17:18:21.972Z]     "findings": [

[2021-05-27T17:18:21.972Z]       {

[2021-05-27T17:18:21.972Z]         "paths": [

[2021-05-27T17:18:21.972Z]           "@storybook/addon-essentials>@storybook/addon-docs>@storybook/builder-webpack4>react-dev-utils>browserslist", 

[2021-05-27T17:18:21.972Z]           "@storybook/addon-essentials>@storybook/addon-docs>@storybook/core>@storybook/core-server>@storybook/builder-webpack4>react-dev-utils>browserslist", 

[2021-05-27T17:18:21.972Z]           "@storybook/react>@storybook/core>@storybook/core-server>@storybook/builder-webpack4>react-dev-utils>browserslist", 

[2021-05-27T17:18:21.972Z]           "@storybook/react>react-dev-utils>browserslist"

[2021-05-27T17:18:21.972Z]         ], 

[2021-05-27T17:18:21.972Z]         "version": "4.14.2"

[2021-05-27T17:18:21.972Z]       }

[2021-05-27T17:18:21.972Z]     ], 

[2021-05-27T17:18:21.972Z]     "title": "Regular Expression Denial of Service", 

[2021-05-27T17:18:21.972Z]     "deleted": null, 

[2021-05-27T17:18:21.972Z]     "overview": "The package `browserslist` from 4.0.0 and before 4.16.5 are vulnerable to Regular Expression Denial of Service (ReDoS) during parsing of queries.", 

[2021-05-27T17:18:21.972Z]     "created": "2021-05-24T19:56:39.062Z", 

[2021-05-27T17:18:21.972Z]     "metadata": {

[2021-05-27T17:18:21.972Z]       "affected_components": "", 

[2021-05-27T17:18:21.972Z]       "module_type": "", 

[2021-05-27T17:18:21.972Z]       "exploitability": 5

[2021-05-27T17:18:21.972Z]     }, 

[2021-05-27T17:18:21.972Z]     "access": "public", 

[2021-05-27T17:18:21.972Z]     "reported_by": {

[2021-05-27T17:18:21.972Z]       "link": "", 

[2021-05-27T17:18:21.972Z]       "name": "Anonymous", 

[2021-05-27T17:18:21.972Z]       "email": ""

[2021-05-27T17:18:21.972Z]     }, 

[2021-05-27T17:18:21.972Z]     "vulnerable_versions": ">=4.0.0 <4.16.5", 

[2021-05-27T17:18:21.972Z]     "references": "- [CVE](https://nvd.nist.gov/vuln/detail/CVE-2021-23364)\n- [GitHub Advisory](https://github.com/advisories/GHSA-w8qv-6jwh-64r5)\n", 

[2021-05-27T17:18:21.972Z]     "url": "https://npmjs.com/advisories/1747", 

[2021-05-27T17:18:21.972Z]     "recommendation": "Upgrade to version 4.16.5 or later", 

[2021-05-27T17:18:21.972Z]     "found_by": {

[2021-05-27T17:18:21.972Z]       "link": "", 

[2021-05-27T17:18:21.972Z]       "name": "Anonymous", 

[2021-05-27T17:18:21.972Z]       "email": ""

[2021-05-27T17:18:21.972Z]     }, 

[2021-05-27T17:18:21.972Z]     "module_name": "browserslist", 

[2021-05-27T17:18:21.972Z]     "patched_versions": ">=4.16.5", 

[2021-05-27T17:18:21.972Z]     "cwe": "CWE-400", 

[2021-05-27T17:18:21.972Z]     "id": 1747, 

[2021-05-27T17:18:21.972Z]     "cves": [

[2021-05-27T17:18:21.972Z]       "CVE-2021-23364"

[2021-05-27T17:18:21.972Z]     ]

[2021-05-27T17:18:21.972Z]   }

[2021-05-27T17:18:21.972Z] }
```

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 1]
```diff
@@ -753,6 +758,22 @@ std::string BraveReferralsService::FormatExtraHeaders(
...
 
+#if defined(OS_ANDROID)
+void BraveReferralsService::InitAndroidReferrer() {
+  android_brave_referrer::InitReferrerCallback init_referrer_callback =
+      base::BindOnce(&BraveReferralsService::OnAndroidBraveReferrerReady,
+                     weak_factory_.GetWeakPtr());
+  android_brave_referrer_.InitReferrer(std::move(init_referrer_callback));
+}
+
+void BraveReferralsService::OnAndroidBraveReferrerReady() {
```
Comment (simonhong) [inline components/brave_referrals/browser/brave_referrals_service.cc:769]: How about adding `DCHECK_CURRENTLY_ON(content::BrowserThread::UI);` here ?

**Review by simonhong — APPROVED** [review]

**Comment by samartnik** [comment]
> @samartnik Or how about calling `BraveReferralsService::Start()` when code is ready instead of pausing in the middle of service initialization?

@simonhong not sure about that since we need that referrer only at the end of initialization and only once. So I would leave it as is, unless you have concerns about current implementation.

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 1 above)
Comment (samartnik) [reply components/brave_referrals/browser/brave_referrals_service.cc:769]: Added

**Comment by simonhong** [comment]
> > @samartnik Or how about calling `BraveReferralsService::Start()` when code is ready instead of pausing in the middle of service initialization?
> 
> @simonhong not sure about that since we need that referrer only at the end of initialization and only once. So I would leave it as is, unless you have concerns about current implementation.

IMO, Calling `Start()` by `OnReferralReady()` is much simpler and straightforward instead of pausing initialization step and waiting. Isn't it? `Start()` is also need to be called only once.

**Comment by samartnik** [comment]
> > > @samartnik Or how about calling `BraveReferralsService::Start()` when code is ready instead of pausing in the middle of service initialization?
> > 
> > 
> > @simonhong not sure about that since we need that referrer only at the end of initialization and only once. So I would leave it as is, unless you have concerns about current implementation.
> 
> IMO, Calling `Start()` by `OnReferralReady()` is much simpler and straightforward instead of pausing initialization step and waiting. Isn't it? `Start()` is also need to be called only once.

`InitAndroidReferrer` needs to be called once per app lifetime, not once per each app start. And we do not pause initialization, referrer initialization happens in a separate non-UI thread.

**Comment by simonhong** [comment]
> > > > @samartnik Or how about calling `BraveReferralsService::Start()` when code is ready instead of pausing in the middle of service initialization?
> > > 
> > > 
> > > @simonhong not sure about that since we need that referrer only at the end of initialization and only once. So I would leave it as is, unless you have concerns about current implementation.
> > 
> > 
> > IMO, Calling `Start()` by `OnReferralReady()` is much simpler and straightforward instead of pausing initialization step and waiting. Isn't it? `Start()` is also need to be called only once.
> 
> `InitAndroidReferrer` needs to be called once per app lifetime, not once per each app start. And we do not pause initialization, referrer initialization happens in a separate non-UI thread.

Ah, Sorry, I was confused and I thought `InitAndroidReferrer()` is called once per app start.
You're right. it only need to be called once. and `Start()` should be called at every app start.
Got it. Let's go as-is.

**Pull request merged**